### ISSUE TITLE
Deal with many users per CWS

### ISSUE THREAD

**Issue by stefano-maggiolo** [cms-dev/cms#375]
At the moment we use a pool_size of 20 with max_overflow of 10 (cms/db/**init**.py). This means that any process will have 20 connection to the DB "ready to use" and requests for more than 30 simultaneous connection will be dropped.

This is a problem only for CWS - almost all page requests require a connection, and so (depending on the rate of page requests per user) if there are more than a certain number of users per CWS, this will lead to errors like those outlined in the troubleshooting documentation page.

"Solving" this could mean several things:
- can we raise substantially the max_overflow number? This possibly means that we would allow users to DoS our DB, but as most queries per page requests are light, I doubt this would be a problem.
- would it make sense to adapt pool_size and max_overflow to the number of users per CWS?
- can we give a more meaningful error message when this happens, with directions on how to solve it?
- can we inform admins before this happens that the ratio users/CWS is too high?
- can we maybe plug this into a safety checker tool that informs admins of common configuration errors?

**Comment by stefano-maggiolo** [comment]
Related change in 4a31176ac92e9daacbf34cb74c86731ee2062942 : raises from 10 to 20 max_overflow, and gives an indication to admins when this might be a problem.

**Comment by lerks** [comment]
My thoughts on the subject: 

As there is only one DB, starting an additional instance of CWS effectively raises the number of simultaneous active connections to the DB by up to pool_size + max_overflow. Thus having many CWS instances with small pools or only one instance with a huge pool is probably identical from the point of view of the DB. Thus pool_size is actually acting just as a limit of the number of concurrent requests that a CWS is willing to handle. This number should loosely be the resources of the machine CWS is running on divided by the resources required to handle a request. My guess is this number is very large, as the only non I/O work of CWS is basically rendering templates, and that shouldn't be that heavy.

In a "safe" environment (e.g. a training camp or the practice session of a medium-large competition) we should try to run a single CWS instance with a huge pool (1000?) to see if that causes any issues.

**Comment by andreyv** [comment]
If I am not mistaken, there seems to be some misconception here. The database connection pool size does not limit the number of simultaneous accepted connections. If there are many accepted connections, they will be just placed in the database pool's queue [1][2] and wait to be processed.

Increasing the pool size has the effect of using more file descriptors and, more importantly, more RAM. For each connection, PostgreSQL creates a new process that consumes memory. I initially started investigating this because our CMS setup (running two ContestWebServers) spawned up to ~70 PostgreSQL processes which consumed a fair amount of RAM.

In my understanding, increasing the pool size also does not mean that more SQL queries will be processed in parallel. In the end this will be limited by the number of available CPUs, and anything higher than that will be queued all the same, but now at the OS level and less efficiently.

I used the Apache Benchmark tool [3] to test CWS under load. Using the following command line:

    ab -n 4000 -c 100 -l -k -C login='<my login cookie>' http://10.0.0.3:8888/tasks/Test/submissions

(100 parallel requests and 4000 in total) I tested two configurations, current CMS master and default SQLAlchemy pool parameters. I found that in both configurations CWS was able to process all parallel requests and that there was no statistical difference in performance. However, my VM's peak memory usage was 534 MB vs 483 MB — a 50 MB difference for one CWS. Detailed results follow (best over multiple runs):

Master:
```
Concurrency Level:      100
Time taken for tests:   89.198 seconds
Complete requests:      4000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      36917514 bytes
HTML transferred:       35055186 bytes
Requests per second:    44.84 [#/sec] (mean)
Time per request:       2229.949 [ms] (mean)
Time per request:       22.299 [ms] (mean, across all concurrent requests)
Transfer rate:          404.18 [Kbytes/sec] received
```

Default SQLAlchemy pool settings [4]:
```
Concurrency Level:      100
Time taken for tests:   87.892 seconds
Complete requests:      4000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      36917586 bytes
HTML transferred:       35055154 bytes
Requests per second:    45.51 [#/sec] (mean)
Time per request:       2197.306 [ms] (mean)
Time per request:       21.973 [ms] (mean, across all concurrent requests)
Transfer rate:          410.19 [Kbytes/sec] received
```

Moreover, there is a vast difference in average response time. Master:
```
Percentage of the requests served within a certain time (ms)
  50%   2179
  66%   2329
  75%   2379
  80%   2407
  90%   2491
  95%   3517
  98%   3878
  99%   4909
 100%  10429 (longest request)
```
Default SQLAlchemy settings:
```
Percentage of the requests served within a certain time (ms)
  50%    362
  66%   2144
  75%   2349
  80%   3998
  90%   6105
  95%   8038
  98%  11947
  99%  14329
 100%  26482 (longest request)
```

While with the default pool settings some requests took longer, 50% of the requests were served in under ~370 ms instead of ~2200 ms for CMS master.

Based on this data, I suggest to return the default SQLAlchemy settings and prevent an explosion of PostgreSQL process count under load. Tuning is likely better done with the `QueuePool`'s `timeout` parameter [5], although the default value is high already.

I can create a pull request if there are no objections.

Regarding this note in the documentation:
> Slightly different, but related, is another issue: CMS may be unable to create new connections to the database because its pool is exhausted. In this case you probably want to modify the `pool_size` argument in `cms/db/__init__.py` or try to spread your users over more instances of ContestWebServer.

How can I reproduce this? With the above setup 100 parallel requests are handled fine.

[1] http://docs.sqlalchemy.org/en/latest/core/pooling.html#sqlalchemy.pool.QueuePool
[2] http://docs.sqlalchemy.org/en/latest/core/engines.html#sqlalchemy.create_engine.params.pool_size
[3] https://httpd.apache.org/docs/2.4/programs/ab.html
[4] https://gist.github.com/andreyv/52d595c15cd5a0b4d38245b3907d2614
[5] http://docs.sqlalchemy.org/en/latest/core/pooling.html#sqlalchemy.pool.QueuePool.params.timeout

Commit references: https://github.com/cms-dev/cms/commit/710ddb5fda68ef99de86975ea20ca3610e61a8bc https://github.com/cms-dev/cms/commit/0b28e6a5e44f3309f7cc6fc85b18947407d4999a

**Comment by wil93** [comment]
> How can I reproduce this? With the above setup 100 parallel requests are handled fine.

Mmm... try with more than 100 parallel requests. I remember this happening during a contest with about 1000 parallel connections (I think we actually had like 4 or 5 running ContestWebServer processes at that time).

**Comment by andreyv** [comment]
I tried with 200 parallel requests, which is pushing the limit of what one CWS can handle (average response time was 4.7 seconds).

About 0.8% of all requests resulted in a timeout. After setting `pool_timeout=60` in the `create_engine()` call and re-running the test, all requests were processed. This is the resulting response time table:
```
Percentage of the requests served within a certain time (ms)
  50%    405
  66%   4696
  75%   5115
  80%   9160
  90%  14237
  95%  19070
  98%  28013
  99%  32509
 100%  54562 (longest request)
```

**Comment by wil93** [comment]
Yes, at this point I think it's reasonable to assume that my issue was mainly caused by `pool_timeout`. So I, for one, agree on reverting to the SQLAlchemy default values. I would also propose to increase `pool_timeout` significantly (is there a way to remove the timeout completely? if there is, is there a reason not to remove the timeout?)

**Comment by andreyv** [comment]
Well, each queued connection consumes resources, so with a huge or infinite timeout it may be possible to DoS the server. I think the timeout should be in the range between 30 and 60 seconds based on the above data. As you can see from average response times above, if requests start failing, the server is already overloaded, and the correct solution is to add more servers instead of increasing the timeout.

**Comment by andreyv** [comment]
I committed the possible changes in andreyv/cms@fecd28c for comments and discussion.

**Comment by andreyv** [comment]
For better clarity I added the changes as a pull request.

Recently we tested these settings in a ~250 user contest with two ContestWebServers, memory usage was stable and I didn't notice delays in the web interface. Only in the first few seconds there was a CPU load peak, which is expected.

**Comment by stefano-maggiolo** [comment]
Thanks for the investigation work and clarity of explanation!

**Comment by stefano-maggiolo** [comment]
Are we satisfied enough to close the original issue?
